### PR TITLE
Order the page models by the order of candidates

### LINKED ISSUE #740: Weird error in symfony routing (only with curl)
This one is a really crazy bug:

- Create two website roots in different languages (e.g. "de" and "en")
- Create a page with identical alias in both languages (e.g. "main")
- Create a subpage in both languages with different aliases (e.g. "main/sub-de" and "main/sub-en")

The page tree should look like this:
```
de [root]
- main (main.html)
- - sub-de (main/sub-de.html)
en [root]
- main (main.html)
- - sub-en (main/sub-en.html)
```
You can access all four pages in your browser, but...
if you try to call them via curl (e.g. https://onlinecurl.com/), one of the subpages returns a 404.
(It depends on which main-page was created first.)

I guess contao/symfony "selects" the wrong "main"-page and cannot find the subpage in the wrong language. I don't know why it works in the browser.

Can anyone confirm?
I would love to debug this. Can someone give me a hint where to look?

### PR DESCRIPTION
**Pull request by leofeyer**
This PR fixes #740.

### CODE REVIEW

File: core-bundle/src/Routing/RouteProvider.php [hunk 1]
```diff
@@ -488,9 +491,15 @@ private function findPages(array $candidates): array
             $conditions[] = 'tl_page.alias IN ('.implode(',', $aliases).')';
         }
 
+        $options = [];
+
+        if (!empty($set)) {
+            $options['order'] = Database::getInstance()->findInSet('alias', $set);
```
Comment (leofeyer) [inline core-bundle/src/Routing/RouteProvider.php]: I don't really like that we have to work with the `Database` object here. Maybe we can use https://github.com/beberlei/DoctrineExtensions here?

**Review by aschempp — CHANGES_REQUESTED** [review]
Can we please have a functional test (first)? Like

1. commit failing functional test
2. make PR with failing functional test
3. rebase/pick this fix on top of a new branch based on top of the failing test
4. create PR second with functional test and fix

This way we can see that this fix is actually correct 🤷‍♂

File: core-bundle/src/Routing/RouteProvider.php (hunk 1 above)
Comment (aschempp) [inline core-bundle/src/Routing/RouteProvider.php]: We should not use the old database connection in the new code

Comment (leofeyer) [reply core-bundle/src/Routing/RouteProvider.php]: I do not like this, either. We could use `FIND_IN_SET()` directly instead?

Comment (leofeyer) [reply core-bundle/src/Routing/RouteProvider.php]: Changed in 07c629278061b4576a55aff83745d7a446a696ae.

File: core-bundle/src/Routing/RouteProvider.php [hunk 2]
```diff
@@ -486,11 +487,12 @@ private function findPages(array $candidates): array
 
         if (!empty($aliases)) {
             $conditions[] = 'tl_page.alias IN ('.implode(',', $aliases).')';
+            $options['order'] = sprintf('FIELD(alias, %s)', implode(',', $aliases));
```
Comment (aschempp) [inline core-bundle/src/Routing/RouteProvider.php]: is this SQL injection safe (because the aliases are taken from the URL)?

Comment (leofeyer) [reply core-bundle/src/Routing/RouteProvider.php]: I have simplified the method in b889f36dc3a610750b68587fdfdfe5d66f7c327d so it works with IDs and aliases and does not need `FIELD()` or `FIND_IN_SET()` at all. Please re-check.

**Review by ausi — APPROVED** [review]

**Comment by xchs** [comment]
Thank you @leofeyer

**Pull request merged**